docs/how_to/s3_integration.md

```diff
@@ -0,0 +1,99 @@
+
+An S3 protocol URL (like `s3://bucket-name/training-data`) can be provided as a parameter for any data iterator that takes a file path as input. For example,
+
+```
+data_iter = mx.io.ImageRecordIter(
+    path_imgrec="s3://bucket-name/training-data/caltech_train.rec",
+    data_shape=(3, 227, 227),
+    batch_size=4,
+    resize=256)
+```
+Following is detailed instructions on how to use data from S3 for training.
+
+## Step 1: Build MXNet with S3 integration enabled
+
+Follow instructions [here](http://mxnet.io/get_started/install.html) to install MXNet from source with the following additional steps to enable S3 integration.
+
+1. Install `libcurl4-openssl-dev` and `libssl-dev` before building MXNet. These packages are required to read/write from AWS S3.
+2. Append `USE_S3 = 1` to `config.mk` before building MXNet.
+    ```
+    echo "USE_S3=1" >> config.mk
+    ```
+
+## Step 2: Configure S3 authentication tokens
+
+MXNet requires the S3 environment variables `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` to be set. [Here](https://aws.amazon.com/blogs/security/wheres-my-secret-access-key/) are instructions to get the access keys from AWS console.
+
+```
+export AWS_ACCESS_KEY_ID=<your-access-key-id>
+AWS_SECRET_ACCESS_KEY=<your-secret-access-key>
+```
+
+## Step 3: Upload data to S3
+
+There are several ways to upload data to S3. One easy way is to use the AWS command line utility. For example, the following `sync` command will recursively copy contents from a local directory to a directory in S3.
+
+```
+aws s3 sync ./training-data s3://bucket-name/training-data
+```
+
+## Step 4: Train with data from S3
+
+Once the data is in S3, it is very straightforward to use it from MXNet. Any data iterator that can read/write data from a local drive can also read/write data from S3.
+
+Let's modify an existing example code in MXNet repository to read data from S3 instead of local disk. [`mxnet/tests/python/train/test_conv.py`](https://github.com/dmlc/mxnet/blob/master/tests/python/train/test_conv.py) trains a convolutional network using MNIST data from local disk. We'll do the following change to read the data from S3 instead.
+
+```
+~/mxnet$ sed -i -- 's/data\//s3:\/\/bucket-name\/training-data\//g' ./tests/python/train/test_conv.py
+
+~/mxnet$ git diff ./tests/python/train/test_conv.py
+diff --git a/tests/python/train/test_conv.py b/tests/python/train/test_conv.py
+index 039790e..66a60ce 100644
+--- a/tests/python/train/test_conv.py
++++ b/tests/python/train/test_conv.py
+@@ -39,14 +39,14 @@ def get_iters():
+
+     batch_size = 100
+     train_dataiter = mx.io.MNISTIter(
+-            image="data/train-images-idx3-ubyte",
+-            label="data/train-labels-idx1-ubyte",
++            image="s3://bucket-name/training-data/train-images-idx3-ubyte",
```

Comment:
Yes, they have to use their own bucket name unless we want to host MNIST data on S3 with read access to everyone.